Add end2end tests.

AlbertMitjans opened the pull request:
This PR copies the end2end tests from `qgqs` into `qiboconnection`. It also updates the `tests` workflow to run the end2end tests for each new PR.

### Next steps:
1. Remove tests from qgqs and use the tests from qiboconnection for the daily end2end workflow.

### Notes:
- For some reason CodeCov's `project` check fails (it says that the overall coverage of the repo decreased...). I don't know why, because I haven't touched the unit tests nor the source code.

JavierSab commented:
uhm... at qiboconnection PRs, are we requiring also e2e? Are changes at backend triggering this WF? If so... can this lead to some kind of loop for changes that modify the contract?

AlbertMitjans commented:
> uhm... at qiboconnection PRs, are we requiring also e2e? Are changes at backend triggering this WF? If so... can this lead to some kind of loop for changes that modify the contract?

No, QGQS has its own workflow, which clones the code from `qiboconnection` and runs the end2end tests. Thus we won't have any loop.

Here's a summary of what I want to do:
1. Move the end2end tests to `qiboconnection` (done in this PR). 
2. Run end2end tests in each `qiboconnection` PR (done in this PR).
3. Remove end2end tests in QGQS and use the ones from `qiboconnection`. Done in [Move end2end tests to qiboconnection](https://github.com/qilimanjaro-tech/qili-global-quantum-service/pull/310#top)
4. Modify the end2end workflow in QGQS to run matrix tests:

| QGQS version | `qiboconnection` version | Results | If tests fail... |
|--------|--------|--------|--------|
| stable | minimum version required (e.g. 0.14.4) | Should always **pass** | Increase minimum version ASAP |
| stable | intermediate versions (e.g. 0.14.6 - 0.15.0) | Should always **pass** | Increase minimum version ASAP |
| stable | last version (e.g. 0.15.0) | Should always **pass** | Release new `qiboconnection` version + increase minimum version ASAP |
| stable | latest (main branch) | Might **fail** | Fix before new `qiboconnection` release |
| latest | minimum version required (e.g. 0.14.4) | Might **fail**  | Increase minimum version before QGQS release |
| latest | intermediate versions (e.g. 0.14.6 - 0.15.0) | Might **fail**  | Increase minimum version before QGQS release |
| latest | last version (e.g. 0.15.0) | Might **fail** () | Release new `qiboconnection` version and increase minimum version before QGQS release |
| latest | latest (main branch) | Might **fail** | Fix changes before release |

ziiiki commented on tests/end2end/test_sanity_checks.py:
```diff
@@ -0,0 +1,29 @@
+    assert isinstance(initial_job_count, int)
+
+
+@pytest.mark.order(-1)
```
-1 means this test is executed the last one?

AlbertMitjans replied:
Yes! Check here: https://pytest-order.readthedocs.io/en/stable/usage.html#order-using-ordinals

ziiiki commented:
I think some tests are missing. I have just run the end2end and obtained 184 test cases (--slow), while in the slack report there are 190

AlbertMitjans commented:
> I think some tests are missing. I have just run the end2end and obtained 184 test cases (--slow), while in the slack report there are 190

Very nice catch!! 😍 

I believe this is because I removed useless parametrisations. In the `test_user_roles.py` file there are some tests that parametrise over users and over devices, but the devices are never used. See example:

```python
@pytest.mark.parametrize(
    "device, user_role",
    [(device, user_role) for user_role in list_user_roles() for device in get_devices_listing_params(user_role)],
)
@pytest.mark.slow
def test_cannot_list_saved_experiments(device: Device, experiment_dict: dict, user_role: UserRole):
    """Test user roles that aren't allowed to list saved experiments can do it -- e.g bsc
    Args:
        api: api instance to call the server with
    """

    user_api = get_user_cannot_post_and_list_experiment_api(user_role=user_role)

    with pytest.raises(requests.exceptions.HTTPError):
        user_api._list_saved_experiments()
```

In this PR pylint complained (hence the importance of pylint checks, again 😆 ) and I removed it. 

Can you try changing the parametrisation of the two tests inside this file and compare the amount of tests?

Here are the 2 corrected tests:

```python
@pytest.mark.parametrize("user_role", [user_role for user_role in list_user_roles()])
@pytest.mark.slow
def test_can_list_saved_experiments(user_role: UserRole):
    """Test user roles that are allowed to list saved experiments can do it -- e.g admin and qili-user
    Args:
        api: api instance to call the server with
    """

    user_api = get_user_can_post_and_list_experiment_api(user_role=user_role)

    assert isinstance(user_api._list_saved_experiments().dataframe, pd.DataFrame)


@pytest.mark.parametrize("user_role", [user_role for user_role in list_user_roles()])
@pytest.mark.slow
def test_cannot_list_saved_experiments(user_role: UserRole):
    """Test user roles that aren't allowed to list saved experiments can do it -- e.g bsc
    Args:
        api: api instance to call the server with
    """

    user_api = get_user_cannot_post_and_list_experiment_api(user_role=user_role)

    with pytest.raises(requests.exceptions.HTTPError):
        user_api._list_saved_experiments()
```

ziiiki commented:
> > I think some tests are missing. I have just run the end2end and obtained 184 test cases (--slow), while in the slack report there are 190
> 
> Very nice catch!! 😍
> 
> I believe this is because I removed useless parametrisations. In the `test_user_roles.py` file there are some tests that parametrise over users and over devices, but the devices are never used. See example:
> 
> ```python
> @pytest.mark.parametrize(
>     "device, user_role",
>     [(device, user_role) for user_role in list_user_roles() for device in get_devices_listing_params(user_role)],
> )
> @pytest.mark.slow
> def test_cannot_list_saved_experiments(device: Device, experiment_dict: dict, user_role: UserRole):
>     """Test user roles that aren't allowed to list saved experiments can do it -- e.g bsc
>     Args:
>         api: api instance to call the server with
>     """
> 
>     user_api = get_user_cannot_post_and_list_experiment_api(user_role=user_role)
> 
>     with pytest.raises(requests.exceptions.HTTPError):
>         user_api._list_saved_experiments()
> ```
> 
> In this PR pylint complained (hence the importance of pylint checks, again 😆 ) and I removed it.
> 
> Can you try changing the parametrisation of the two tests inside this file and compare the amount of tests?
> 
> Here are the 2 corrected tests:
> 
> ```python
> @pytest.mark.parametrize("user_role", [user_role for user_role in list_user_roles()])
> @pytest.mark.slow
> def test_can_list_saved_experiments(user_role: UserRole):
>     """Test user roles that are allowed to list saved experiments can do it -- e.g admin and qili-user
>     Args:
>         api: api instance to call the server with
>     """
> 
>     user_api = get_user_can_post_and_list_experiment_api(user_role=user_role)
> 
>     assert isinstance(user_api._list_saved_experiments().dataframe, pd.DataFrame)
> 
> 
> @pytest.mark.parametrize("user_role", [user_role for user_role in list_user_roles()])
> @pytest.mark.slow
> def test_cannot_list_saved_experiments(user_role: UserRole):
>     """Test user roles that aren't allowed to list saved experiments can do it -- e.g bsc
>     Args:
>         api: api instance to call the server with
>     """
> 
>     user_api = get_user_cannot_post_and_list_experiment_api(user_role=user_role)
> 
>     with pytest.raises(requests.exceptions.HTTPError):
>         user_api._list_saved_experiments()
> ```

Apart from that, there's a couple of excepts in t`est_end2end.py `that are not here and are in qgqs main

ziiiki commented:
> > > I think some tests are missing. I have just run the end2end and obtained 184 test cases (--slow), while in the slack report there are 190
> > 
> > 
> > Very nice catch!! 😍
> > I believe this is because I removed useless parametrisations. In the `test_user_roles.py` file there are some tests that parametrise over users and over devices, but the devices are never used. See example:
> > ```python
> > @pytest.mark.parametrize(
> >     "device, user_role",
> >     [(device, user_role) for user_role in list_user_roles() for device in get_devices_listing_params(user_role)],
> > )
> > @pytest.mark.slow
> > def test_cannot_list_saved_experiments(device: Device, experiment_dict: dict, user_role: UserRole):
> >     """Test user roles that aren't allowed to list saved experiments can do it -- e.g bsc
> >     Args:
> >         api: api instance to call the server with
> >     """
> > 
> >     user_api = get_user_cannot_post_and_list_experiment_api(user_role=user_role)
> > 
> >     with pytest.raises(requests.exceptions.HTTPError):
> >         user_api._list_saved_experiments()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In this PR pylint complained (hence the importance of pylint checks, again 😆 ) and I removed it.
> > Can you try changing the parametrisation of the two tests inside this file and compare the amount of tests?
> > Here are the 2 corrected tests:
> > ```python
> > @pytest.mark.parametrize("user_role", [user_role for user_role in list_user_roles()])
> > @pytest.mark.slow
> > def test_can_list_saved_experiments(user_role: UserRole):
> >     """Test user roles that are allowed to list saved experiments can do it -- e.g admin and qili-user
> >     Args:
> >         api: api instance to call the server with
> >     """
> > 
> >     user_api = get_user_can_post_and_list_experiment_api(user_role=user_role)
> > 
> >     assert isinstance(user_api._list_saved_experiments().dataframe, pd.DataFrame)
> > 
> > 
> > @pytest.mark.parametrize("user_role", [user_role for user_role in list_user_roles()])
> > @pytest.mark.slow
> > def test_cannot_list_saved_experiments(user_role: UserRole):
> >     """Test user roles that aren't allowed to list saved experiments can do it -- e.g bsc
> >     Args:
> >         api: api instance to call the server with
> >     """
> > 
> >     user_api = get_user_cannot_post_and_list_experiment_api(user_role=user_role)
> > 
> >     with pytest.raises(requests.exceptions.HTTPError):
> >         user_api._list_saved_experiments()
> > ```
> 
> Apart from that, there's a couple of excepts in t`est_end2end.py `that are not here and are in qgqs main

Just run qgqs main tests with the changes you mention above, and obtained 184 cases. So once missing except clauses are added to` test_end2end.py`, good to go!

AlbertMitjans commented:
> So once missing except clauses are added to test_end2end.py, good to go!

What missing except clauses? 🤔
